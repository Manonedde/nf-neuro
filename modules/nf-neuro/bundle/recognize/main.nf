process BUNDLE_RECOGNIZE {
    tag "$meta.id"
    label 'process_high'

    container "${ workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container ?
        'https://scil.usherbrooke.ca/containers/scilus_latest.sif':
        'scilus/scilus:latest' }"

    input:
        tuple val(meta), path(tractograms), path(transform), path(config), path(directory)

    output:
    tuple val(meta), path("*_cleaned.trk")            , emit: bundles
    tuple val(meta), path("*_bundles_mosaic.png")     , emit: mosaic
    tuple val(meta), path("*_bundles_stats.json")     , emit: stats
    path "versions.yml"                               , emit: versions

    when:
    task.ext.when == null || task.ext.when

    script:
    def prefix = task.ext.prefix ?: "${meta.id}"

    // additional script arguments
    def minimal_vote_ratio = task.ext.minimal_vote_ratio ? "--minimal_vote_ratio " + task.ext.minimal_vote_ratio : ""
    def seed = task.ext.seed ? "--seed " + task.ext.seed : ""
    def rbx_processes = task.ext.rbx_processes ? "--processes " + task.ext.rbx_processes : "--processes 1"
    def outlier_alpha = task.ext.outlier_alpha ? "--alpha " + task.ext.outlier_alpha : ""
    def run_qc = task.ext.run_qc ? task.ext.run_qc : false
    """
    mkdir recobundles/
    scil_tractogram_segment_bundles.py ${tractograms} ${config} ${directory}/ ${transform} --inverse --out_dir recobundles/ \
        -v DEBUG $minimal_vote_ratio $seed $rbx_processes

    for bundle_file in recobundles/*.trk; do
        bname=\$(basename \${bundle_file} .trk)
        out_cleaned=${prefix}__\${bname}_cleaned.trk
        scil_bundle_reject_outliers.py \${bundle_file} "\${out_cleaned}" ${outlier_alpha}
    done


    if $run_qc; then
        scil_tractogram_compute_density_map.py recobundles/PYT_l.trk tmp_anat.nii.gz --binary
        scil_viz_bundle_screenshot_mosaic.py tmp_anat.nii.gz *_cleaned.trk --opacity_background 1
        scil_bundle_shape_measures.py *_cleaned.trk --out_json ${prefix}__bundles_stats.json
    fi


    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        scilpy: 2.0.0
    END_VERSIONS
    """

    stub:
    def prefix = task.ext.prefix ?: "${meta.id}"
    """
    scil_tractogram_segment_bundles.py -h
    scil_bundle_reject_outliers.py -h

    # dummy output for single bundle
    touch ${prefix}__AF_L_cleaned.trk

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        scilpy: 2.0.0
    END_VERSIONS
    """
}
