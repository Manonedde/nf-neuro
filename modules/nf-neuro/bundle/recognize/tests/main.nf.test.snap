{
    "bundle - recognize - params": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        [
                            "test__bundle_1_cleaned.trk:md5,4f5cb88d05f175d6b55509d0ad307ab3",
                            "test__bundle_2_cleaned.trk:md5,58b4c1381603d72ad6e72cc705b4606e",
                            "test__bundle_3_cleaned.trk:md5,8f24a8c534b6e2386bcc755bd8f0c4b9",
                            "test__bundle_4_cleaned.trk:md5,eafac0796051d6ada4bfc5a5c576c536",
                            "test__bundle_5_cleaned.trk:md5,0d512bcaaf49325415a1f4860a167f5e"
                        ]
                    ]
                ],
                "1": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test__bundles_mosaic_mqc.png:md5,0c801fbb54350275f472025b4aebe128"
                    ]
                ],
                "2": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test__bundles_stats_mqc.json:md5,8c18b46194ec557b34bcf4136ee801ce"
                    ]
                ],
                "3": [
                    "versions.yml:md5,083946242b74b84b09704b8be661eee2"
                ],
                "bundles": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        [
                            "test__bundle_1_cleaned.trk:md5,4f5cb88d05f175d6b55509d0ad307ab3",
                            "test__bundle_2_cleaned.trk:md5,58b4c1381603d72ad6e72cc705b4606e",
                            "test__bundle_3_cleaned.trk:md5,8f24a8c534b6e2386bcc755bd8f0c4b9",
                            "test__bundle_4_cleaned.trk:md5,eafac0796051d6ada4bfc5a5c576c536",
                            "test__bundle_5_cleaned.trk:md5,0d512bcaaf49325415a1f4860a167f5e"
                        ]
                    ]
                ],
                "global_mqc": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test__bundles_stats_mqc.json:md5,8c18b46194ec557b34bcf4136ee801ce"
                    ]
                ],
                "mqc": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test__bundles_mosaic_mqc.png:md5,0c801fbb54350275f472025b4aebe128"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,083946242b74b84b09704b8be661eee2"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.0",
            "nextflow": "24.10.4"
        },
        "timestamp": "2025-02-25T18:00:24.361429655"
    },
    "bundle - recognize - base": {
        "content": [
            {
                "0": [
                    
                ],
                "1": [
                    
                ],
                "2": [
                    
                ],
                "3": [
                    
                ],
                "bundles": [
                    
                ],
                "global_mqc": [
                    
                ],
                "mqc": [
                    
                ],
                "versions": [
                    
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.0",
            "nextflow": "24.10.4"
        },
        "timestamp": "2025-02-25T15:00:01.833495184"
    }
}